File: Code/1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start
import collections
from collections import defaultdict

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        my_dict = defaultdict(lambda: [])
        for (index, value) in enumerate(nums):
            my_dict[value].append(index)
        nums.sort()
        print(nums)
        head_index = 0
        tail_index = len(nums) - 1
        head = nums[head_index]
        tail = nums[tail_index]
        if head_index == tail_index:
            return []
        while head_index != tail_index:
            sum1 = head + tail
            print(sum1)
            if sum1 == target:
                index1 = my_dict[head].pop()
                index2 = my_dict[tail].pop()
                res = [index1, index2] if index1 < index2 else [index2, index1]
                return res

            if head_index == tail_index:
                return []
            elif sum1 > target:
                tail_index -= 1
            elif sum1 < target:
                head_index += 1
            
            head = nums[head_index]
            tail = nums[tail_index]
        return []


        
        
# @lc code=end

File: Code/1.two-sum.java
/*
 * @lc app=leetcode id=1 lang=java
 *
 * [1] Two Sum
 */

// @lc code=start
class Solution {
    public int[] twoSum(int[] nums, int target) {
        //Method using set
        Map<Integer,Integer> map = new HashMap();
        for(int i = 0;i<nums.length;i++){
            map.put(nums[i], i);
        }
        for(int i = 0;i<nums.length;i++){
            if(map.containsKey(target-nums[i]) && !map.get(target-nums[i]).equals(i))
                return new int[]{i, map.get(target-nums[i])};
        }
        return null;
    }
}
// @lc code=end

